File: 113-path-sum-ii/path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    void dfs(TreeNode *root, int sum, int curr, vector<int> temp, vector<vector<int>> &ans) {
        if(root == NULL) return;
        curr += root->val;
        temp.push_back(root->val);
        
        if(curr == sum && root->left == NULL && root->right == NULL) 
            ans.push_back(temp);
        dfs(root->left, sum, curr, temp, ans);
        dfs(root->right, sum, curr, temp, ans);
    }
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> ans;
        vector<int> temp;
        dfs(root, targetSum, 0, temp, ans);
        return ans;
    }
};

File: 113-path-sum-ii/path-sum-ii.java

class Solution {
    private static void dfs(TreeNode root, int sum, int curr, ArrayList<Integer> temp, List<List<Integer>> ans) {
        if (root == null) return;

        curr += root.val;
        temp.add(root.val);

        if (curr == sum && root.left == null && root.right == null) {
            ans.add(new ArrayList<>(temp));
        }

        dfs(root.left, sum, curr, temp, ans);
        dfs(root.right, sum, curr, temp, ans);

        // Backtrack: remove the last element before returning from the current call
        temp.remove(temp.size() - 1);
    }

    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> ans = new ArrayList<>();
        ArrayList<Integer> temp = new ArrayList<>();
        dfs(root, targetSum, 0, temp, ans);
        return ans;
    }
}
